### DISCUSSION TITLE
Need Help Changing Proxy Without Leaving the Browser

### DISCUSSION THREAD

**Discussion by Yanni-Benji** [seleniumbase/SeleniumBase#2615] [Q&A] [answered]
I am currently facing an issue with my proxy, and I would appreciate your assistance in resolving it. Here’s my situation:

I want to modify my proxy settings without leaving my web browser. Specifically, I’d like my proxy to change automatically for each request I make—whether it’s accessing a new page or refreshing an existing one. I am using an API, and the proxy format is as follows: “http://{USERNAME}:{PASSWORD}@{PROXY_DNS}.”

If anyone has advice or instructions on how to achieve this, I would be very grateful. Thank you in advance for your help!

```
from seleniumbase import SB
username = "USERNAME"
password = "PASSWORD"
PROXY_DNS = "00.000.000.000:0000"
urlToGet = "http://ip-api.com/json"
proxy = "{}:{}@{}".format(username, password, PROXY_DNS)
print(proxy)
with SB( proxy=proxy) as sb:

    sb.get(urlToGet)
###### I want here to change the proxy and execute again sb.get(urlToGet)
```

**Comment by mdmintz** [comment, marked as answer]
Partial duplicate of https://github.com/seleniumbase/SeleniumBase/issues/2100#issuecomment-1716091343 (That was for `BaseCase`, but you're using the `SB()` format):

To change the proxy of an active browser in the middle of a test: (`SB()` format):
* Use SeleniumBase with Wire Mode enabled (``wire=True`` / ``--wire``)
* Use the `sb.set_wire_proxy()` method:

```python
sb.set_wire_proxy("SERVER:PORT")
sb.set_wire_proxy("socks5://SERVER:PORT")
sb.set_wire_proxy("USERNAME:PASSWORD@SERVER:PORT")
```

#### (Note that Wire Mode cannot be used together with UC Mode.)

**Reply by Yanni-Benji** [reply]
I’ve tried this code, but it doesn’t change the IP address every time I access the website. 
```
from seleniumbase import SB
from time import sleep
username = "USERNAME"
password = "PASSWORD"
PROXY_DNS = "00.000.000.000:0000"
urlToGet = "http://ip-api.com/json"
proxies = "{}:{}@{}".format(username, password, PROXY_DNS)
print(proxies)
with SB(proxy = proxies, wire=True) as sb:
    
    for i in range(20):
        sb.get(urlToGet)
        print(sb.get_page_source())
        sb.driver.set_wire_proxy(proxies)
        sleep(10)
```
When I try to use the requests package, it works for each of my requests with the following line of code:
`r = requests.get(urlToGet, proxies=proxy)`
So the problem doesn’t come from the proxy API.

**Reply by mdmintz** [reply]
You want `sb.set_wire_proxy(proxy)`. (`sb.driver` doesn't have that method.)

**Reply by Yanni-Benji** [reply]
I just forget to remove `.driver` in this code(I just tried, and i forgot to take it off in the response and it doesn't work with `sb.get_wire_proxy(proxies)`)

**Reply by mdmintz** [reply]
See https://github.com/wkeeling/selenium-wire/blob/da1b675fe2cc6dae2e3bb959c1ce95eb1c41830b/README.rst?plain=1#L707

The SeleniumBase wire methods call selenium-wire directly. Eg:

```python
def set_wire_proxy(string):
    """Examples:
    set_wire_proxy("SERVER:PORT")
    set_wire_proxy("socks5://SERVER:PORT")
    set_wire_proxy("USERNAME:PASSWORD@SERVER:PORT")
    """
    the_http = "http"
    the_https = "https"
    if string.startswith("socks4://"):
        the_http = "socks4"
        the_https = "socks4"
    elif string.startswith("socks5://"):
        the_http = "socks5"
        the_https = "socks5"
    string = string.split("//")[-1]
    driver.proxy = {
        "http": "%s://%s" % (the_http, string),
        "https": "%s://%s" % (the_https, string),
        "no_proxy": "localhost,127.0.0.1",
    }
```

If you have an issue with the SeleniumBase version try the selenium-wire API directly. If you have any issues with that, you'll have to open a ticket with them directly.

**Reply by Yanni-Benji** [reply]
I will see that.  Thank you very much sir.
